R/convert_output_types.R

```diff
@@ -0,0 +1,162 @@
+#' terminal output types, and the resulting output will only contain the terminal
+#' output types. See details for supported conversions.
+#'
+#' @param model_out_tbl an object of class `model_out_tbl` containing predictions
+#'    with a single, unique value in the `output_type` column.
+#' @param terminal_output_type character vector of the desired output type(s)
+#'   following transformation. May contain any of the following output types:
+#'   `"mean"`, `"median"`, `"quantile"`
+#' @param terminal_output_type_id (generally) a named list indicating the desired
+#'   output type IDs for each new output type. Can be one of three possible data
+#'   types, for the following output types:
+#'   - `NA` for `"mean"` and `"median"` (no associated output type ID)
+#'   - a numeric vector of probability levels for `"quantile"`
+#'
+#'   If only one new output type is requested, the singular list item may be supplied
+#'   directly. Defaults to `NA`. See the examples for an illustration of both cases.
+#'
+#' @details Currently, only `"sample"` can be converted to `"mean"`, `"median"`, or `"quantile"`
+#'
+#' @examples
+#' # We illustrate the conversion between output types using normal distributions
+#' ex_quantiles <- c(0.25, 0.5, 0.75)
+#' model_out_tbl <- expand.grid(
+#'   stringsAsFactors = FALSE,
+#'   group1 = c(1,2),
+#'   model_id = "A",
+#'   output_type = "sample",
+#'   output_type_id = 1:100
+#' ) |>
+#' dplyr::mutate(value = rnorm(200, mean = group1))
+#'
+#' # Multiple output type conversions in one call
+#' convert_output_type(model_out_tbl, terminal_output_type = c("quantile", "median"),
+#'    terminal_output_type_id = list("quantile" = ex_quantiles, "median" = NA))
+#'
+#' # Single output type conversion
+#' convert_output_type(model_out_tbl, terminal_output_type = "quantile",
+#'    terminal_output_type_id = ex_quantiles)
+#'
+#' @return object of class `model_out_tbl` containing (only) predictions of the
+#'   terminal output_type(s) for each unique combination of task IDs for each model
+#'
+#' @export
+#' @importFrom rlang .data
+convert_output_type <- function(model_out_tbl, terminal_output_type,
+                                terminal_output_type_id = NA) {
+  # validations
+  initial_output_type <- unique(model_out_tbl$output_type)
+  task_id_cols <- subset_task_id_names(names(model_out_tbl))
+  validate_conversion_inputs(
+    initial_output_type, terminal_output_type,
+    terminal_output_type_id
+  )
+
+  # pre-processing
+  grouped_model_outputs <- dplyr::group_by(model_out_tbl, dplyr::across(dplyr::all_of(c("model_id", task_id_cols))))
+  # main calculations
+  purrr::map(terminal_output_type, .f = function(terminal_output_type_tmp) {
+    if (is.list(terminal_output_type_id)) {
+      terminal_output_type_id_tmp <- terminal_output_type_id[[terminal_output_type_tmp]]
+    } else {
+      terminal_output_type_id_tmp <- terminal_output_type_id
+    }
```

Comment:
Instead of creating a new variable with a `tmp` suffix, it would be better to overwrite the `terminal_output_type_id` variable since its scope is protected within the function context. 

```suggestion
    if (is.list(terminal_output_type_id)) {
      terminal_output_type_id <- terminal_output_type_id[[terminal_output_type_tmp]]
    }
```